### PR TITLE
Resumption supported on multi-threaded server

### PR DESCRIPTION
**Pull request by somasun**
Major changes in this PR are 
- `FrameTransport` is an interface which `FrameTransportImpl` and `ScheduledFrameTransport` implement.  `FrameTransportImpl` is the old `FrameTransport` (just renaming here).
- `RSocketStateMachine` uses the `FrameTransport` interface to interact with the transport
- `ScheduledFrameProcessor` has been introduced.
- If the resumed connection is on a different EventBase, we use `ScheduledFrameTransport` and `ScheduledFrameProcessor` to ensure the `RSocketStateMachine` continues to live on the old EventBase, but communicate with the transport on the new EventBase.
- Else, there is no thread hopping and regular FrameTransport and FrameProcessor (RSocketStateMachine) is used.

In the future, if/when we get rid of FrameTransport, we have to just give a different implementation of the FrameTransport interface.  The current PR is not bound to FrameTransport in anyway. 

I tested by increasing the number of threads in the resumption-server (both unittest and examples) and verified that the resumption succeeds when resumed on different EventBase.  AsyncServerSocket round-robins among the available worker threads, so this was pretty deterministic.   I will follow up with stress testing resumption

### CODE REVIEW

File: rsocket/framing/ScheduledFrameProcessor.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#include "rsocket/framing/ScheduledFrameProcessor.h"
+
+#include <folly/ExceptionWrapper.h>
+#include <folly/io/IOBuf.h>
+
+namespace rsocket {
+
+void ScheduledFrameProcessor::processFrame(
+    std::unique_ptr<folly::IOBuf> ioBuf) {
+  // ScheduledFrameProcessor could get destroyed before the scheduled
```
Comment (lexs) [inline rsocket/framing/ScheduledFrameProcessor.cpp]: You can get around this by submitting a copy of it only `~ScheduledFrameProcessor` gets called. Due to EventBases strict ordering you'll guarantee destruction doesn't happen until all callbacks have been processed.

This way we can avoid incrementing/decrementing the ref-count for every frame.

File: rsocket/framing/ScheduledFrameProcessor.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  // event gets executed.  Preserve a copy of inner frameProcessor_
+  evb_->runInEventBaseThread(
+      [ frameProcessor = frameProcessor_, ioBuf = std::move(ioBuf) ]() mutable {
+        frameProcessor->processFrame(std::move(ioBuf));
+      });
+}
+
+void ScheduledFrameProcessor::onTerminal(folly::exception_wrapper ex) {
+  // ScheduledFrameProcessor could get destroyed before the scheduled
+  // event gets executed.  Preserve a copy of inner frameProcessor_
```
Comment (lexs) [inline rsocket/framing/ScheduledFrameProcessor.cpp]: ^ same

File: rsocket/framing/ScheduledFrameTransport.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    auto scheduledFP = std::make_shared<ScheduledFrameProcessor>(
+        std::move(fp), stateMachineEvb_);
+    frameTransport_->setFrameProcessor(std::move(scheduledFP));
+  });
+}
+
+void ScheduledFrameTransport::outputFrameOrDrop(
+    std::unique_ptr<folly::IOBuf> ioBuf) {
+  // ScheduledFrameTransport could get destroyed before the scheduled
+  // event gets executed.  Preserve a copy of inner frameTransport_
```
Comment (lexs) [inline rsocket/framing/ScheduledFrameTransport.cpp]: Same reasoning as other class.

File: rsocket/framing/ScheduledFrameProcessor.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include <folly/io/async/EventBase.h>
+
+#include "rsocket/framing/FrameProcessor.h"
+
+namespace rsocket {
+
+class ScheduledFrameProcessor : public FrameProcessor {
```
Comment (lexs) [inline rsocket/framing/ScheduledFrameProcessor.h]: Can you add some doc what this solves? Other class too.

File: rsocket/RSocketServer.cpp [hunk 5]
```diff
@@ -161,11 +162,32 @@ void RSocketServer::onRSocketResume(
...
   CHECK(serverState);
+  auto* eventBase = folly::EventBaseManager::get()->getExistingEventBase();
+  VLOG(2) << "Resuming client on " << eventBase->getName();
   if (!serverState->eventBase_.isInEventBaseThread()) {
-    throw RSocketException("Trying to RESUME in a different worker thread");
+    // If the resumed connection is on a different EventBase, then use
+    // ScheduledFrameTransport and ScheduledFrameProcessor to ensure the
+    // RSocketStateMachine continues to live on the same EventBase and the
+    // IO happens in the new EventBase
+    auto scheduledFT = yarpl::make_ref<ScheduledFrameTransport>(
```
Comment (phoad) [inline rsocket/RSocketServer.cpp]: Can we move the creation of scheduledFT into the lambda?

File: rsocket/framing/FrameTransport.h [hunk 6]
```diff
@@ -1,68 +1,20 @@
...
-
-  void closeImpl(folly::exception_wrapper);
-
-  std::shared_ptr<FrameProcessor> frameProcessor_;
-  std::shared_ptr<DuplexConnection> connection_;
+namespace rsocket {
 
-  yarpl::Reference<DuplexConnection::Subscriber> connectionOutput_;
-  yarpl::Reference<yarpl::flowable::Subscription> connectionInputSub_;
+class FrameTransport : public virtual yarpl::Refcounted {
```
Comment (phoad) [inline rsocket/framing/FrameTransport.h]: nit: do we need an interface to be ref counted?

File: rsocket/framing/ScheduledFrameProcessor.cpp (hunk 1 above)
Comment (phoad) [reply rsocket/framing/ScheduledFrameProcessor.cpp]: 💯

File: rsocket/framing/ScheduledFrameProcessor.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ public:
+  ScheduledFrameProcessor(
+      std::shared_ptr<FrameProcessor> fp,
```
Comment (phoad) [inline rsocket/framing/ScheduledFrameProcessor.h]: Why not a unique_ptr? I would prefer the object that a class decorates to take the ownership of it.

File: rsocket/framing/FrameTransport.h [hunk 8]
```diff
@@ -1,68 +1,20 @@
...
+#include <folly/io/async/EventBase.h>
 
-  bool isClosed() const {
-    return !connection_;
-  }
+#include "rsocket/framing/FrameProcessor.h"
 
- private:
-  void connect();
+using folly::EventBase;
```
Comment (alexmalyshev) [inline rsocket/framing/FrameTransport.h]: Drive-by nit: This shouldn't be in the header, nor is it used.

File: rsocket/RSocketServer.cpp (hunk 5 above)
Comment (somasun) [reply rsocket/RSocketServer.cpp]: No strong preference either way.  I found this easier to read.

File: rsocket/framing/FrameTransport.h (hunk 6 above)
Comment (somasun) [reply rsocket/framing/FrameTransport.h]: FrameTransport is passed around in rsocket lib as yarpl::ref

File: rsocket/framing/ScheduledFrameProcessor.cpp [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+#include <folly/io/IOBuf.h>
+
+namespace rsocket {
+
+ScheduledFrameProcessor::~ScheduledFrameProcessor() {
+  // ScheduledFrameProcessor could get destroyed before the scheduled events of
+  // the class get executed.  So preserve a copy of inner frameProcessor_ in
+  // the event queue.  Since the events in the queue are executed in order we
+  // are guaranteed to have the frameProcessor_ during all the other events
+  // which were added before the destruction of this class.
```
Comment (alexmalyshev) [inline rsocket/framing/ScheduledFrameProcessor.cpp]: :+1: for comment

File: rsocket/transports/tcp/TcpConnectionAcceptor.cpp [hunk 10]
```diff
@@ -72,8 +73,10 @@ void TcpConnectionAcceptor::start(OnDuplexConnectionAccept onAccept) {
   callbacks_.reserve(options_.threads);
   for (size_t i = 0; i < options_.threads; ++i) {
     callbacks_.push_back(std::make_unique<SocketCallback>(onAccept_));
-    callbacks_[i]->eventBase()->runInEventBaseThread(
-        [] { folly::setThreadName("TcpConnectionAcceptor.Worker"); });
+    callbacks_[i]->eventBase()->runInEventBaseThread([i] {
+      folly::EventBaseManager::get()->getEventBase()->setName(
+          folly::sformat("TCPAcceptor.Worker.{}", i));
```
Comment (alexmalyshev) [inline rsocket/transports/tcp/TcpConnectionAcceptor.cpp]: :+1:

File: rsocket/framing/ScheduledFrameTransport.h [hunk 11]
```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <folly/io/async/EventBase.h>
+
+#include "rsocket/framing/FrameTransportImpl.h"
+#include "rsocket/framing/ScheduledFrameProcessor.h"
+
+using folly::EventBase;
```
Comment (alexmalyshev) [inline rsocket/framing/ScheduledFrameTransport.h]: Nit: We shouldn't use `using' in headers.

File: rsocket/framing/FrameTransport.h [hunk 12]
```diff
@@ -1,68 +1,18 @@
...
-
-  void closeImpl(folly::exception_wrapper);
-
-  std::shared_ptr<FrameProcessor> frameProcessor_;
-  std::shared_ptr<DuplexConnection> connection_;
+namespace rsocket {
 
-  yarpl::Reference<DuplexConnection::Subscriber> connectionOutput_;
-  yarpl::Reference<yarpl::flowable::Subscription> connectionInputSub_;
+class FrameTransport : public virtual yarpl::Refcounted {
```
Comment (alexmalyshev) [inline rsocket/framing/FrameTransport.h]: Nit: Drop a comment here saying to read `FrameTransportImpl`'s comments 😛 .

File: rsocket/framing/FrameTransport.h [hunk 13]
```diff
@@ -1,68 +1,18 @@
...
-
-  /// Cancel the input, complete the output, and close the underlying
-  /// connection.
-  void close();
+#include "yarpl/Refcounted.h"
 
-  /// Cancel the input, error the output, and close the underlying connection.
-  /// This must be closed with a non-empty exception_wrapper.
-  void closeWithError(folly::exception_wrapper);
+#include <folly/io/async/EventBase.h>
```
Comment (alexmalyshev) [inline rsocket/framing/FrameTransport.h]: We're not using `EventBase` in this header, can drop this.

**Comment by alexmalyshev** [comment]
> Why aren't there any changes to WarmResumptionTest? IIRC WarmResumptionTest.FailedResumption1 runs a server with 2 threads, and assumes the first resumption attempt will fail. That should no longer be the case right?

Talked IRL, but it's cause the server in those tests isn't resumable.  Ignore me.

**Comment by somasun** [comment]
We will use SwappableEventBase or better approaches for the longer term.  This is to get things running in the short term.
